spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/TcpConnectionInterceptorSupport.java

```diff
@@ -101,7 +101,7 @@ public void registerSender(TcpSender sender) {
 
 	@Override
 	public void registerSenders(List<TcpSender> sendersToRegister) {
-		this.theConnection.registerSenders(sendersToRegister);
+		this.theConnection.registerSenders(sendersToRegister, this);
```

Comment:
It is indeed confusing that interceptor abstraction implements a connection one.
So, we have such a strange code flow...

I wish @garyrussell will come back to us one day to revise this architecture 😄 

Nevertheless would you mind, @gigermocas , to double check what we have so far with the `TcpConnectionSupport.close()`:
```
public void close() {
		for (TcpSender sender : this.senders) {
			sender.removeDeadConnection(this);
		}
...
```
After this fix it doesn't look like we are going to clean up properly...

I'll debug it locally meanwhile, too

Reply:
>I wish @garyrussell will come back to us one day to revise this architecture 😄

Yes, this is a bit of a mess; dates back to very early code and has caused headaches since then.

Maybe in 6.0 @artembilan if you want to raise an issue.

Reply:
Issue created: https://github.com/spring-projects/spring-integration/issues/3515, but it doesn't mean we have to make it in `6.0`. Probably one day in the future when it causes more headaches 😉